### PR TITLE
Arc 1659 get facebook business manager data into arc marts

### PR DESCRIPTION
**Pull request by ryantimjohn**
# Pull Request for dbt-arc-functions

- [x] Open the pull request as a draft so that you can run tests against the code and inspect the files in the Files view of the Pull Request interface in Github.

- [x] Once you're satisfied, convert the draft into an open Pull Request by clicking "Ready for Review" near the bottom of this page. You should not make any changes to the code except in dire emergency once you've converted from draft, so make sure you're really satisfied.

- [x] Add two of the principal contributors to this PR for review. Currently, the principal contributors are @dmbluestate , @Frydafly , and @ryantimjohn.

- [x] We've divided our PR templates into Macros, Utils and Docs PR requests. Delete the ones you're not using.
### Macros (delete if not using)

- [x] Link Jira tickets (or Github Issues) to PR here. Make sure that the Jira ticket has a good description as to why this PR is necessary and the business problem that this PR is solving. Moreover, if there have been any discussions of this ticket in Slack, make sure they're linked in the Jira ticket:

https://bluestate.atlassian.net/browse/ARC-1659

- [x] Make sure this is named using Jira extension so that naming convention is consistent. If it's not, that's okay. Go to the Jira ticket, click into the ticket, then on the right under Details, find Development, then click the Create branch button. Copy the Branch Name from there. Read here [how to rename a branch in Github](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch) to rename this branch. In the future, create branches from Jira.

- [x] Make sure the Macro is documented by running the notebook: `create_docs.ipynb` util in this repo. From there, you'll have to either fill in descriptions by hand OR use our `fill_in_descriptions_using_openai.ipynb` to fill in descriptions using OpenAI's ChatGPT. Always make sure to check over ChatGPT's column descriptions, as they're likely not perfect.

- [x] Test branch on a development branch of an existing client (ideally the one that raised the bug). Do this by changing revision of the package to the branch name in `packages.yml` file in dbt State client below.

https://github.com/bsd/arc-comicrelief-dbt/pull/1

- [x] If new macro files were created or updated, re-run `create_or_update_standard_models.py` for the client dbt project, replacing models and dependencies

- [x] check in dbt that client package successfully runs `dbt deps`, `dbt compile`, and `dbt run`; screenshot this page

Screenshot 2023-11-07 at 5 05 01 PM

- [x] query the resulting staging model in bigquery (or by previewing mart in dbt cloud, or your IDE); screenshot this

Screenshot 2023-11-07 at 5 36 29 PM

### CODE REVIEW

File: documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+models:
+  - name: stg_stitch_facebook_paidmedia_campaigns
+    description: Model for storing data related to Facebook paid media campaigns.
+    columns:
+      - name: campaign_id
+        data_type: string
+        description: The ID of the campaign.
+      - name: message_id
+        data_type: string
+        description: "The ID of the ad.\n\nReferenced in\n\nads.id\nads.targeting.id\n\
```
Comment (Frydafly) [inline documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.yml:12]: not that readible in yml form 

`description: |` 

also an option, but i don't feel strongly about this either way

File: documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_clicks_daily_rollup.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+---
+version: 2
+models:
+  - name: stg_stitch_facebook_paidmedia_clicks_daily_rollup
+    description: This model represents the daily rollup of paid media clicks from
+      the Facebook advertising platform.
+    columns:
+      - name: message_id
+        data_type: string
+        description: The ID of the ad. It is referenced in ads.id, ads.targeting.id,
```
Comment (Frydafly) [inline documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_clicks_daily_rollup.yml:10]: flagging multiline,  might need `|` or `>`

Comment (Frydafly) [reply documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_clicks_daily_rollup.yml:10]: or if it's the same language referenced in the other yml files, then maybe a `.md` file and `docs` block reference?

File: documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+---
+version: 2
+models:
+  - name: stg_stitch_facebook_paidmedia_impressions_daily_rollup
+    description: Impressions daily rollup data for paid media on Facebook
+    columns:
+      - name: message_id
+        data_type: string
+        description: The ID of the ad. Referenced in ads.id, ads.targeting.id, ads_insights.ad_id,
```
Comment (Frydafly) [inline documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.yml:9]: multi line

File: macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{% macro create_stg_stitch_facebook_paidmedia_campaigns(
+    source_name="src_stitch_facebook_paidmedia",
+    source_table="ads_insights"
+) %}
+    select
+        campaign_id as campaign_id,
+        ad_id as message_id,
+        'social' as channel_category,
+        'facebook' as channel,
```
Comment (Frydafly) [inline macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql:9]: possible the value for channel is usually `facebook_ads`-- not sure though, would be good to double check

File: macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{% macro create_stg_stitch_facebook_paidmedia_impressions_daily_rollup(
+    source_name="src_stitch_facebook_paidmedia",
+    source_table="ads_insights"
+) %}
+    select
+        ad_id as message_id,
+        safe_cast(date_start as timestamp) as date_timestamp,
+        impressions as total_impressions,
+        reach as unique_impressions
```
Comment (Frydafly) [inline macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.sql:9]: just checking on the reach = unique_impressions?

**Review by Frydafly — APPROVED** [review]
just got a few comments + questions! looks good to me after that

File: macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.sql (hunk 5 above)
Comment (ryantimjohn) [reply macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_impressions_daily_rollup.sql:9]: yep!

File: macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql (hunk 4 above)
Comment (ryantimjohn) [reply macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql:9]: `facebook_ad` ! TY

Screenshot 2023-11-08 at 11 11 21 AM

File: documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.yml (hunk 1 above)
Comment (ryantimjohn) [reply documentation/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.yml:12]: I tried that and it got corrected to this... don't understand yamlfix

File: macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql (hunk 4 above)
Comment (ryantimjohn) [reply macros/stitch_facebook_paidmedia/staging/stg_stitch_facebook_paidmedia_campaigns.sql]: resolved

**Comment by ryantimjohn** [comment]
@Frydafly understand the concern about multiline but yamlfix corrected from multiline to this so... I'm just going to take yamlfix's edits because checks won't pass otherwise. Will see down the line if there's any way to turn off this not allowing multiline setting 😞

**Review by dmbluestate — COMMENTED** [review]
The screenshot shows that may be duplicates in the mart, and looking at the mart table there are both duplicated from the same ad and the same day, as well as rows with all nulls across all columns.
Screenshot 2023-11-08 at 11 28 44 AM
Screenshot 2023-11-08 at 11 28 54 AM

**Comment by ryantimjohn** [comment]
@dmbluestate bug ticket created! https://bluestate.atlassian.net/browse/ARC-1700

Will take a look at this next week, will see if this is join related or something in the dataset itself!

**Pull request merged**